nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py

```diff
@@ -607,6 +615,289 @@ def forward(
 
         return (packed_result,)
 
+    def _greedy_decode_cuda_optimized_lstm(
+        self,
+        x: torch.Tensor,
+        out_len: torch.Tensor,
+        device: torch.device,
+        max_symbols_per_hyp_factor: float = 0.5,
+        partial_hypotheses: Optional[List[rnnt_utils.Hypothesis]] = None,
+    ):
+        if partial_hypotheses is not None:
+            raise NotImplementedError("`partial_hypotheses` support is not supported")
+
+        with torch.inference_mode():
+            # x: [B, T, D]
+            # out_len: [B]
+            # device: torch.device
+
+            # Initialize list of Hypothesis
+            batchsize = x.shape[0]
+
+            max_symbols_per_hyp = max(int(max_symbols_per_hyp_factor * x.shape[1]), 256)
+            hypotheses = rnnt_utils.CudaHypothesesTransducer(batchsize, max_symbols_per_hyp, x.device)
+
+            # Initialize Hidden state matrix (shared by entire batch)
+            hidden = None
+
+            # Last Label buffer + Last Label without blank buffer
+            # batch level equivalent of the last_label
+            last_label = torch.full([batchsize, 1], fill_value=self._blank_index, dtype=torch.long, device=device)
+
+            # Get max sequence length
+            max_out_len = out_len.max()
+
+            D = x.shape[-1]
+            f = torch.zeros(
+                [batchsize, 1, D], dtype=x.dtype, device=x.device
+            )  # to store features of hyps for all batch-id
+
+            hyp2t = torch.zeros(
+                [batchsize], dtype=torch.long, device=x.device
+            )  # hyp2t[i] represent the time step of i'th hyp in kept_hyps
+
+            hyp2looptimes = torch.zeros(
+                [batchsize], dtype=torch.long, device=x.device
+            )  # to keep track of how many times the decoding stays at the same frame to compare with max_symbols_per_step
+
+            done = torch.zeros([batchsize], dtype=torch.bool, device=x.device)
+            hyp2b = torch.LongTensor([i for i in range(batchsize)]).to(x.device)
+
+            shift = hyp2b
+
+            x2 = torch.reshape(x, [-1, D])
+
+            bos = True
+            while True:
```

Comment:
Could you provide a detailed breakdown of decoding loop and every step inside. Basically please note the overall high level algorithm that is being done here and below. At the high level i do understand it but better to have it written down for future.